lib/router/src/lib.rs

```diff
@@ -72,30 +82,63 @@ where
 {
     recognize: Rec,
     make: Mk,
-    cache: Mutex<Cache<Rec::Target, LoadShed<Mk::Value>>>,
+    cache: Lock<Cache<Rec::Target, LoadShed<Mk::Value>>>,
 }
 
-enum State<Req, Svc>
+struct RouteRequest<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req>,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    Init(Option<Req>, Svc),
-    Called(Svc::Future),
+    request: Option<Req>,
+    target: Option<Rec::Target>,
+    make: Option<Mk>,
+    cache: Lock<Cache<Rec::Target, LoadShed<Mk::Value>>>,
+}
+
+enum State<Req, Rec, Mk>
+where
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req>,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
+{
+    Inserting(RouteRequest<Req, Rec, Mk>),
```

Comment:
Is `Inserting` the best name for this still? We are not necessarily inserting anything in this state...

Perhaps rename these enum states as `Acquire/Call/Respond`?